src/read.jl

```diff
@@ -261,3 +256,236 @@ function read_demand_flexibility(filepath, interval)::DemandFlexibility
+            flexible_bus_idx = [sets.bus_id2idx[x] for x in flexible_bus_id]
+            flexible_bus_num = length(flexible_bus_id)
+
+            # remove bus columns from zone to bus mapping
+            zone_to_bus_shares[:, flexible_bus_idx] .= 0
+
+            # re-normalize the rows to distribute flexiblity among un-specified buses
+            for i in 1:length(zone_list)
+                zone_to_bus_shares[i, :] ./= sum(zone_to_bus_shares[i, :])
+            end
+
+            # check if zone numbers are correct
+            if !all([issubset(i, zone_list) for i in flexible_zone_id])
+                throw(
+                    ErrorException(
+                        "Invalid load zone numeric ID(s) in demand flexibility input files!"
+                    ),
+                )
+            elseif !all([issubset(i, case.busid) for i in flexible_bus_id])
+                throw(
+                    ErrorException(
+                        "Invalid load bus numeric ID(s) in demand flexibility input files!"
+                    ),
+                )
+            end
+
+            # list index of each zone in input file in the sorted list of zones
+            zone_cols_idx = [findfirst(y -> y == x, zone_list) for x in flexible_zone_id]
+
+            # for flex amt, the zone numbers are the total flexibility in the zone
+            if field == "flex_amt_up" || field == "flex_amt_dn"
+                # if input contains bus columns
+                if flexible_bus_num > 0
+                    # find if the flexibility of any bus is also specified in a zone column
+                    bus_cols_zone_id = case.bus_zone[findall(
+                        x -> x in flexible_bus_id, case.busid
+                    )]
+                    bus_cols_zone_idx = [
+                        findfirst(isequal(i), flexible_zone_id) for i in bus_cols_zone_id
+                    ]
+
+                    # substract individual bus columns from zone total flexibility
+                    for i in 1:flexible_bus_num
+                        # substract from total if the zone of this bus is specified in a zone column
+                        if !isnothing(bus_cols_zone_idx[i])
+                            fh[:, zone_columns_idx[bus_cols_zone_idx[i]] + 1] -= fh[
+                                :, bus_columns_idx[i] + 1
+                            ]
+                        end
+                    end
+
+                    # check if flexibility in any zone is less than the sum of buses
+                    for i in 1:flexible_zone_num
+                        if any(x -> x < 0, fh[:, zone_columns_idx[flexible_zone_num] + 1])
+                            throw(
+                                ErrorException(
+                                    "Input ERROR: Total zone flexibility less than sum of bus
+                       flexibility for zone " * string(flexible_zone_id[i]),
+                                ),
```

Comment:
This passed the linter? The formatting on the string in the `ErrorException` seems a little odd.

Reply:
This is created by lint.. Before I run the formatter it didn't look like this.

Reply:
Thanks for clarifying!